sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java

```diff
@@ -372,60 +375,115 @@ private InputGetterImpl(Expression input, Schema inputSchema) {
-          field =
+          value =
               nullOr(
-                  field,
+                  value,
                   Expressions.convert_(
                       Expressions.divide(millisField, Expressions.constant(MILLIS_PER_DAY)),
                       int.class));
         } else if (!logicalId.equals(CharType.IDENTIFIER)) {
           throw new IllegalArgumentException(
-              "Unknown LogicalType " + fromType.getLogicalType().getIdentifier());
+              "Unknown LogicalType " + type.getLogicalType().getIdentifier());
         }
-      } else if (CalciteUtils.isDateTimeType(fromType)) {
-        field = nullOr(field, Expressions.call(field, "getMillis"));
-      } else if (fromType.getTypeName().isCompositeType()
-          || (fromType.getTypeName().isCollectionType()
-              && fromType.getCollectionElementType().getTypeName().isCompositeType())) {
-        field =
-            Expressions.condition(
-                Expressions.equal(field, Expressions.constant(null)),
-                Expressions.constant(null),
-                Expressions.call(WrappedList.class, "of", field));
-      } else if (fromType.getTypeName() == TypeName.BYTES) {
-        field =
-            Expressions.condition(
-                Expressions.equal(field, Expressions.constant(null)),
-                Expressions.constant(null),
-                Expressions.new_(ByteString.class, field));
+      } else if (CalciteUtils.isDateTimeType(type)) {
+        return nullOr(value, Expressions.call(value, "getMillis"));
+      } else if (type.getTypeName().isCompositeType()) {
+        return nullOr(value, row(value, type.getRowSchema()));
+      } else if (type.getTypeName().isCollectionType()) {
+        return nullOr(value, list(value, type.getCollectionElementType()));
+      } else if (type.getTypeName() == TypeName.BYTES) {
+        return nullOr(
+            value, Expressions.new_(ByteString.class, Types.castIfNecessary(byte[].class, value)));
+      }
+
+      return value;
+    }
+
+    private static Expression list(Expression input, FieldType elementType) {
+      ParameterExpression value = Expressions.parameter(Object.class);
+
+      BlockBuilder block = new BlockBuilder();
+      block.add(value(value, elementType));
+
+      return Expressions.new_(
+          WrappedList.class,
+          ImmutableList.of(Types.castIfNecessary(List.class, input)),
+          ImmutableList.<MemberDeclaration>of(
+              Expressions.methodDecl(
+                  Modifier.PUBLIC,
+                  Object.class,
+                  "value",
+                  ImmutableList.of(value),
+                  block.toBlock())));
+    }
+
+    private static Expression row(Expression input, Schema schema) {
+      ParameterExpression row = Expressions.parameter(Row.class);
+      ParameterExpression index = Expressions.parameter(int.class);
+      BlockBuilder body = new BlockBuilder(/* optimizing= */ false);
```

Comment:
Why does optimizing need to be set to false here and below?

Reply:
Good question. Honestly, I don't know how optimizations work in linq4j, but I expect we wouldn't need any, and we would rather have predictable code tree for easier debugging. Enabling and understanding optimizations would be something to look into later. For instance, now we would generate code like `(new ByteString((byte[])value).getBytes())` that isn't going to be optimized. I guess we can fix that by extending code optimization if needed.